dependencydiff/dependencydiff.go

```diff
@@ -0,0 +1,119 @@
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/ossf/scorecard/v4/checks"
+	"github.com/ossf/scorecard/v4/clients"
+	"github.com/ossf/scorecard/v4/clients/githubrepo"
+	"github.com/ossf/scorecard/v4/clients/githubrepo/roundtripper"
+	"github.com/ossf/scorecard/v4/pkg"
+)
+
+// GetDependencyDiffResults gets dependency changes between two given code commits BASE and HEAD
+// along with the Scorecard check results of the dependencies, and returns a slice of DependencyCheckResult.
+func GetDependencyDiffResults(ownerName, repoName, baseSHA, headSHA string) ([]pkg.DependencyCheckResult, error) {
+	ctx := context.Background()
+	// Fetch dependency diffs using the GitHub Dependency Review API.
+	deps, err := FetchDependencyDiffData(ctx, ownerName, repoName, baseSHA, headSHA)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching dependency changes: %w", err)
+	}
+
+	ghRepo, err := githubrepo.MakeGithubRepo(path.Join(ownerName, repoName))
+	if err != nil {
+		return nil, fmt.Errorf("error creating the github repo: %w", err)
+	}
+	ghRepoClient := githubrepo.CreateGithubRepoClient(ctx, nil)
+	ossFuzzRepoClient, err := githubrepo.CreateOssFuzzRepoClient(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating the oss fuzz repo client: %w", err)
+	}
+	vulnsClient := clients.DefaultVulnerabilitiesClient()
+
+	results := []pkg.DependencyCheckResult{}
+	for _, d := range deps {
+		if d.SourceRepository != nil {
+			if *d.SourceRepository == "" {
+				continue
+			}
+			// Running scorecard on this dependency and fetch the result.
+			// TODO: use the Scorecare REST API to retrieve the Scorecard result statelessly.
+			scorecardResult, err := pkg.RunScorecards(
+				ctx,
+				ghRepo,
+				clients.HeadSHA,
+				checks.AllChecks,
+				ghRepoClient,
+				ossFuzzRepoClient,
+				nil,
+				vulnsClient,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("error fetching the scorecard result: %w", err)
+			}
+			depCheckResult := pkg.DependencyCheckResult{
+				PackageURL:       d.PackageURL,
+				SourceRepository: d.SourceRepository,
+				ChangeType:       d.ChangeType,
+				ManifestPath:     d.ManifestPath,
+				Ecosystem:        d.Ecosystem,
+				Version:          d.Version,
+				ScorecardResults: &scorecardResult,
+				Name:             d.Name,
+			}
+			results = append(results, depCheckResult)
+		}
+		// Skip those without source repo urls.
+		// TODO: use the BigQuery dataset to supplement null source repo URLs
+		// so that we can fetch the Scorecard results for them.
+	}
+	return results, nil
+}
+
+// FetchDependencyDiffData fetches the depednency-diffs between the two code commits
+// using the GitHub Dependency Review API, and returns a slice of Dependency.
+func FetchDependencyDiffData(ctx context.Context, owner, repo, base, head string) ([]Dependency, error) {
+	reqURL, err := url.Parse("https://api.github.com")
+	if err != nil {
+		return nil, fmt.Errorf("error parsing the url: %w", err)
+	}
+	reqURL.Path = path.Join(
```

Comment:
will this fail if someone run scorecard on a Windows machine? Since there will be `\\` instead of `/`?
Maybe that's a problem with other APIs we have !?
There must be a proper http path API that exist for this purpose maybe, https://pkg.go.dev/net/url#QueryEscape?

Reply:
um, i tested it on my personal windows PC and it seems to be fine.

still looking for documents and APIs that can be more properly used for this url construction purpose.

Reply:
Let's not spend too much time on this, it's fine for now. Thanks for checking.